### PR TITLE
Remove puppet3 support - get puppet4 instead!

### PR DESCRIPTION
**Pull request by martbhell**
This is mostly work done by @junousi in use in https://github.com/CSCfi/polte that I have been testing on our normal infrastructure.
Testing:
 - Use this role after deploying a fresh VM to setup a puppetmaster version 4 with collection RPMs. 
 - Use https://github.com/CSCfi/ansible-role-puppetize to puppetize the puppet-agent4.

Merging this makes making changes to the puppet3 code a bit trickier, but one could make another branch for puppet3 from the commit prior to this PR if desired.

 - #CCCP-1278 

Summary of changes:
 - change paths and variables to the new puppetlabs ones
 - pin gems
 - remove passenger code
 - gem install of hiera-eyaml
 - add another scenario where we want to template in puppet.conf
 - add some ansible_managed comments to templates
 - generate types of the puppet_environment - this fixed the "autoload" issues we saw on the first puppet run

### CODE REVIEW

File: templates/puppet.conf
Comment (junousi) [inline templates/puppet.conf]: The puppetmaster doesn't really run an agent. Looking closer at puppet.conf reference, these should probably go to [main] or [master]

**Review by junousi — CHANGES_REQUESTED** [review]

File: defaults/main.yml
Comment (junousi) [inline defaults/main.yml:22]: You tested this with a recent reinstall or two, right? 

I was thinking there might have been some weird corner case where the system-wide hiera-eyaml would be needed alongside the puppetmaster installed one. But I'm probably wrong, especially if it worked on a fresh install :)

Comment (martbhell) [reply defaults/main.yml:22]: Got worried there because the github comment only showed the red bits!
This has been tested with reinstalls and CentOS 7.7 for both environments.
To be honest I'm not sure what is Right or Wrong. It works and I think it's nice to pin the versions :)

File: templates/puppet.conf
Comment (martbhell) [reply templates/puppet.conf]: Hmm.. I'll test. I think it could be nice to be able to run "puppet agent -t" on the puppetmaster too rather than the masterless bit in this role or am I wrong?

File: defaults/main.yml
Comment (junousi) [reply defaults/main.yml:22]: Pinning is fine, this was only concerning the exclusion of hiera-eymaml from this list of OS side gems. Which is fine as well if reinstalls have worked.

File: templates/puppet.conf
Comment (junousi) [reply templates/puppet.conf]: True! I've mostly run the puppet apply thing on the puppetmasters, but OTOH can't think of a reason why we couldn't support running the agent.

Maybe if we were doing the "periodic agent poll + immediate provisioning" thing, then I wouldn't want the puppetmaster agent to immediately apply what has just been merged to production. But in any case that's not a thing right now.

Comment (martbhell) [reply templates/puppet.conf]: Ran some testing and without the [agent] bits then indeed running "puppet agent -t" does not work on the puppetmaster -- I would like to keep it :)

Comment (junousi) [reply templates/puppet.conf]: Yup, resolved

**Comment by junousi** [comment]
Now just need to make travis happy?

**Comment by junousi** [comment]
You could also consider tagging current master ref with a "puppet3-eol" tag or equivalent if we ever need to installa a system into that state for some mysterious reason.

**Comment by martbhell** [comment]
Travis is now green and yay also --check mode works :)

File: tasks/main.yml
Comment (oscarkraemer) [inline tasks/main.yml]: Why is this needed?

File: templates/puppet.conf
Comment (oscarkraemer) [inline templates/puppet.conf]: `ansible_inventory` is probably preferred, probably same is valid for the rest of the `ansible_fqdn` s in the file.

**Review by oscarkraemer — CHANGES_REQUESTED** [review]
- Question about the need of one of tests. If it is needed would `stat` be better suited?
- `ansible_fqdn` -> `ansible_inventory`  
- As a reviewer can I run this against the puppet4 servers to test it? What would be a good command?

File: templates/puppet.conf
Comment (martbhell) [reply templates/puppet.conf]: Good points. Removed the fqdn and changed to the inventory based variable.

File: tasks/main.yml
Comment (martbhell) [reply tasks/main.yml]: The reason why this specific task "check if puppet.conf is OK" was added because when I first started this sprint to look at this the puppetmaster was in an unhappy state (directory there but puppet.conf did not point anywhere) but because the previous check found an existing directory a task below did not template in the puppet.conf

I think I would ask: Why are both of these "check if puppet is configured" needed.

Now. If nothing else seems to manage the puppet.conf on the puppetmaster I think we can actually remove the checks and just always manage these via templates..

Comment (martbhell) [reply tasks/main.yml]: Turns out it's nice to have the old ls etc/dir check to determine if we want to create puppetmaster certificates or not. Adjusted a bit.

**Comment by martbhell** [comment]
> * Question about the need of one of tests. If it is needed would `stat` be better suited?

For that specific test I wanted to check the content of the file. To my knowledge stat does not store that in a variable. But for the other test where we use "ls {etc_dir}" then sure. Just needs different conditionals further down the task file that way. Bit better now.

> * `ansible_fqdn` -> `ansible_inventory`

< inventory_hostname

> * As a reviewer can I run this against the puppet4 servers to test it? What would be a good command?

Sure. You'll need some vars, they can be found in a PR with the same branch name in both of our private ansible-* repos.
To test the puppet4 servers after they have been deployed one can for example run "puppet agent -t".

File: tasks/main.yml
Comment (oscarkraemer) [inline tasks/main.yml]: ansible suggested using the file module to change directory permission.

**Review by oscarkraemer — APPROVED** [review]
Ansible printed a warning about using file-module  instead of the shell module for changing file permissions. I will approve this, if you want to fix that before merging feel free.

File: tasks/main.yml
Comment (martbhell) [reply tasks/main.yml]: Good point. Changed that to file module and used the variable to the directory instead of the hard coded one.

**Comment by martbhell** [comment]
> Ansible printed a warning about using file-module instead of the shell module for changing file permissions. I will approve this, if you want to fix that before merging feel free.

Still has a change request from Jukka so waiting response there before merging :)

**Review by junousi — APPROVED** [review]

**Pull request merged**
